doc/modules/linear_model.rst

```diff
@@ -313,15 +314,75 @@ is correct, i.e. that the data are generated by this model.
 They also tend to break when the problem is badly conditioned
 (more features than samples).
 
-.. figure:: ../auto_examples/linear_model/images/sphx_glr_plot_lasso_model_selection_001.png
-    :target: ../auto_examples/linear_model/plot_lasso_model_selection.html
+.. figure:: ../auto_examples/linear_model/images/sphx_glr_plot_lasso_lars_ic_001.png
+    :target: ../auto_examples/linear_model/plot_lasso_lars_ic.html
     :align: center
     :scale: 50%
 
+Mathematical details
+""""""""""""""""""""
+
+The definition of AIC (and thus BIC) might differ in the literature. In this
+section, we give more information regarding the criterion computed in
+scikit-learn. The AIC criterion is defined as:
+
+.. math::
+    AIC = -2 \log(\hat{L}) + 2 d
+
+where :math:`\hat{L}` is the maximum likelihood of the model and
+:math:`d` is the number of parameters (as well referred to as degrees of
+freedom in the previous section).
+
+For a linear Gaussian model, the maximum log-likelihood is defined as:
+
+.. math::
+    \log(\hat{L}) = - \frac{n}{2} \log(2 \pi) - \frac{n}{2} \ln(\hat{\sigma}^2) - \frac{\sum_{i=1}^{n} (y_i - \hat{y_i})^2}{2\hat{\sigma}^2}
+
+where :math:`\hat{\sigma}^2` is an estimate of the noise variance,
+:math:`y_i` and :math:`\hat{y_i}` are respectively the true and predicted
+targets, and :math:`n` is the number of samples.
+
+Plugging the maximum log-likelihood in the AIC formula yields:
+
+.. math::
+    AIC = n \log(2 \pi \hat{\sigma}^2) + \frac{\sum_{i=1}^{n} (y_i - \hat{y_i})^2}{\hat{\sigma}^2} + 2d
+
+The first term of the above expression is sometimes discarded since it is a
+constant when :math:`\hat{\sigma}^2` is provided. In addition,
+it is sometimes stated that the AIC is equivalent to the :math:`C_p` statistic
+[12]_. However, this is important to note that it is up to some constant
+and factor term.
+
+At last, we mentioned above that :math:`\hat{\sigma}^2` is an estimate of the
+noise variance. In :class:`LassoLarsIC` when the parameter `noise_variance` is
+not provided (default), the noise variance is estimated via the unbiased
+estimator [13]_ defined as:
+
+.. math::
+    \hat{\sigma^2} = \frac{\sum_{i=1}^{n} (y_i - \hat{y_i})^2}{n - p}
+
+where :math:`p` is the number of features and :math:`\hat{y_i}`
+is the predicted target using an ordinary least squares regression. In
+scikit-learn, we use a ridge model with a very small regularization in case
+of ill-conditioned design matrix. Note, that this formula is valid only when
+`n_samples > n_features`.
```

Comment:
Are you sure, using Ridge is fine? Could you test that for one case?
https://arxiv.org/pdf/0712.0881.pdf Theorem 1 states that results are valid only if `X` is full rank (dummy vs one-hot encoding).

Reply:
The safeguard to use an OLS is to check the rank of the matrix before computing the `dof` then?

Reply:
Let's try with plain `LinearRegression`. The coef will be numerically unstable but the estimated variance based on its in-sample predictions should be fine.

We just need to make sure to include tests in with `n_samples ~= n_features` and many collinear features to check that either the results are ok without error messages or that the error messages (if any) make sense.

Reply:
I'm not saying that ridge is false. I'm just asking as the papers do not mention it and use plain OLS. I've no intuition if it's fine in this context.